iavl/iavl_range_proofs.go

```diff
@@ -0,0 +1,419 @@
+type KeyInRangeProof interface {
+	Verify(startKey, endKey, key, value, root []byte) error
+}
+
+// KeyFirstInRangeProof is a proof that a given key is the first in a given range.
+type KeyFirstInRangeProof struct {
+	KeyExistsProof `json:"key_proof"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// String returns a string representation of the proof.
+func (proof *KeyFirstInRangeProof) String() string {
+	return fmt.Sprintf("%#v", proof)
+}
+
+// Verify that the first in range proof is valid.
+func (proof *KeyFirstInRangeProof) Verify(startKey, endKey, key, value []byte, root []byte) (err error) {
+	if key != nil && (bytes.Compare(key, startKey) == -1 || bytes.Compare(key, endKey) == 1) {
+		return ErrInvalidInputs
+	}
+	if proof.Left == nil && proof.Right == nil && proof.PathToKey == nil {
+		return ErrInvalidProof
+	}
+	if err := verifyPaths(proof.Left, proof.Right, startKey, endKey, root); err != nil {
+		return err
+	}
+
+	if proof.PathToKey != nil {
+		if err := proof.KeyExistsProof.Verify(key, value, root); err != nil {
+			return errors.Wrap(err, "failed to verify key exists proof")
+		}
+		// If the key returned is equal to our start key, and we've verified
+		// that it exists, there's nothing else to check.
+		if bytes.Equal(key, startKey) {
+			return
+		}
+		// If the key returned is the smallest in the tree, then it must be
+		// the smallest in the given range too.
+		if proof.PathToKey.isLeftmost() {
+			return
+		}
+		// The start key is in between the left path and the key returned,
+		// and the paths are adjacent. Therefore there is nothing between
+		// the key returned and the start key.
+		if proof.Left != nil && proof.Left.Path.isLeftAdjacentTo(proof.PathToKey) {
+			return
+		}
+	} else if proof.Left != nil && proof.Left.Path.isRightmost() {
+		// No key found. Range starts outside of the right boundary.
+		return
+	} else if proof.Right != nil && proof.Right.Path.isLeftmost() {
+		// No key found. Range ends outside of the left boundary.
+		return
+	} else if proof.Left != nil && proof.Right != nil &&
+		proof.Left.Path.isLeftAdjacentTo(proof.Right.Path) {
+		// No key found. Range is between two existing keys.
+		return
+	}
+
+	return ErrInvalidProof
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// KeyLastInRangeProof is a proof that a given key is the last in a given range.
+type KeyLastInRangeProof struct {
+	KeyExistsProof `json:"key_proof"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// String returns a string representation of the proof.
+func (proof *KeyLastInRangeProof) String() string {
+	// TODO(cloudhead): Needs work.
+	return fmt.Sprintf("%#v", proof)
+}
+
+// Verify that the last in range proof is valid.
+func (proof *KeyLastInRangeProof) Verify(startKey, endKey, key, value []byte, root []byte) (err error) {
```

Comment:
ditto here please, for the comments left for `func (proof *KeyFirstInRangeProof) Verify`